shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Core.hs

```diff
@@ -480,16 +483,26 @@ pickStakeKey keys = vKey . snd <$> QC.elements keys
 -- Note: we need to keep the initial utxo coin sizes large enough so that
 -- when we simulate sequences of transactions, we have enough funds available
 -- to include certificates that require deposits.
-genTxOut :: (ShelleyTest era) => Constants -> [Addr era] -> Gen [TxOut era]
+genTxOut :: forall era. (ShelleyTest era) => Constants -> [Addr era] -> Gen [TxOut era]
 genTxOut Constants {maxGenesisOutputVal, minGenesisOutputVal} addrs = do
-  ys <- genCoinList minGenesisOutputVal maxGenesisOutputVal (length addrs) (length addrs)
-  return (uncurry TxOut <$> zip addrs ys)
+  let ln = length addrs
+  p <- max <$> QC.choose (0, ln) <*> QC.choose (0, ln)
+  cls <- genCoinList minGenesisOutputVal maxGenesisOutputVal p
+  vls <- genValList @era minGenesisOutputVal maxGenesisOutputVal (ln - p)
+  outs <- QC.shuffle ((Val.inject <$> cls) ++ vls)
+  return (uncurry TxOut <$> zip addrs outs)
 
 -- | Generates a list of 'Coin' values of length between 'lower' and 'upper'
 -- and with values between 'minCoin' and 'maxCoin'.
-genCoinList :: Integer -> Integer -> Int -> Int -> Gen [Coin]
-genCoinList minCoin maxCoin lower upper = do
-  len <- QC.choose (lower, upper)
+genValList :: (ShelleyTest era) => Integer -> Integer -> Int -> Gen [Core.Value era]
```

Comment:
I don't think this is the right way to go. This pretends to be a generator of arbitrary `Value era`, but it's only doing so by disguising coins as arbitrary values - they will be zero for any other currency.

I would have thought the better approach would be to keep the generator as specific to coin, but allow e.g. `genTxOut` above to take the value generator as an argument. Then in shelley-ma, we can pass a different generator which creates `Value` with a full range?

Reply:
This is not making 0 of all other tokens. It is making an arbitrary Value (using whatever arbitrary generator we we have to the specific Value for the era), then removing the coins from that, and adding coins generated separately, using the CONSTANTS values (so that the resulting coin distribution is as expected)

Reply:
Ah, okay! My apologies for failing to understand!